hs_core/xmlparser.py

```diff
@@ -0,0 +1,61 @@
+import base64
+import os.path
+from StringIO import StringIO
+from xml.sax import xmlreader, saxutils
+from xml.sax.expatreader import ExpatParser
+
+from django.conf import settings
+
+
+XML_CACHE = os.path.join(settings.PROJECT_ROOT, '_xmlcache')
+
+
+class CatalogedExpatParser(ExpatParser):
+    def external_entity_ref(self, context, base, sysid, pubid):
+        if not self._external_ges:
+            return 1
+
+        source = self._ent_handler.resolveEntity(pubid, sysid)
+        source = saxutils.prepare_input_source(source,
+                                               self._source.getSystemId() or
+                                               "")
+
+        # If an entry does not exist in the xml cache, create it.
+        filepath = os.path.join(XML_CACHE, base64.urlsafe_b64encode(pubid))
+        if not os.path.isfile(filepath):
+            with open(filepath, 'w') as f:
+                contents = source.getByteStream().read()
+                source.setByteStream(StringIO(contents))
+                f.write(contents)
```

Comment:
The documentation is pretty cryptic:  https://docs.python.org/2/library/xml.sax.handler.html#xml.sax.handler.EntityResolver.resolveEntity

To a large extent, I just had to read the code:

https://github.com/python/cpython/blob/2.7/Lib/xml/sax/__init__.py
https://github.com/python/cpython/blob/2.7/Lib/xml/sax/expatreader.py#L390-L410
https://github.com/python/cpython/blob/2.7/Lib/xml/sax/handler.py#L223-L234
https://github.com/python/cpython/blob/2.7/Lib/xml/sax/saxutils.py#L310-L353

Also, the checked in DTD file was pulled down by this code, so it definitely does what it is supposed to.